doc/manual/methods.rst

```diff
@@ -440,6 +440,198 @@ can also constrain type parameters of methods::
+    +(a, b) = +(promote(a, b)...)
+    # once the elements have the same type, they can be added via Intrinsic calls
+    +(a::T, b::T) = Core.add(a, b)
+
+
+4. Trait-based dispatch
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that identifies classes of types.
+The example above glossed over this detail by avoiding a discussion
+of the implementation of ``map`` and ``promote``.
+
+When iterating over a matrix, such as in the implementation of ``map``,
+one important question is what order to use to traverse the data.
+When ``AbstractArray`` subtypes implement the ``linearindexing`` trait,
+other functions such as ``map`` can dispatch on this information to pick the best algorithm.
+This means that each subtypes does not need to implement a custom version of ``map``,
+since the generic definitions will enable the system to select the fastest version.
+
+::
+
+    map(a::AbstractArray, b::AbstractArray) = map(Base.linearindexing(a, b), a, b)
+    map(::LinearSlow, a::AbstractArray, b::AbstractArray) = # generic implementation
+    map(::LinearFast, a::AbstractArray, b::AbstractArray) = # linear-fast implementation
+
+This trait-based mechanism is also present in the ``promote`` mechanism used by ``+``.
+It is using the types to compute the optimal common type for computing the operation.
+This makes it possible to reduce the problem of implementing every function for every pair of possible type arguments,
+to the much smaller problem of implementing a conversion operation from each type to a common type,
+plus a table of preferred pair-wise promotion rules.
+
+
+5. Output-type Computation
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The discussion of trait-based promotion provides a segue into our next design pattern:
+computing the output element type for a matrix operation.
+
+For implementing primitive operations, such as addition,
+we use the ``promote_type`` function to compute the desired output type.
+(As before, we saw this at work in the ``promote`` call in the call to ``+``).
+
+For more complex functions on matrices, it may be necessary to compute the expected return
+type for a more complex sequence of operations.
+This is often performed by the following sequence of operations:
+
+1. Write a small function ``op`` that expresses the set of operations performed by the kernel of the algorithm.
+2. Compute the element type ``R`` of the result matrix as ``promote_op(op, argument_types...)``,
+   where ``argument_types`` is computed from ``eltype`` applied to each input array.
+3. Build the output matrix as ``similar(R, dims)``, where ``dims`` is the desired dimensions of the output array.
+
+For a more specific example, a generic matrix multiply pseudo-code might look like::
+
+    function matmul(a::AbstractMatrix, b::AbstractMatrix)
+        op = (ai, bi) -> ai * bi + ai * bi
+
+        ## this fails because it assumes `one(eltype(a))` is constructable:
```

Comment:
Rather than "fails", better say "may fail for some types" or "is not completely correct".